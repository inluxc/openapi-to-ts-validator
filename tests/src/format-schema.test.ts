import path from "node:path";
import fs from "node:fs";
import { generate } from "openapi-typescript-validator";
import Ajv from "ajv";

describe("format-schema - compile based", () => {
  const name = "format";
  const generatedDir = path.join(__dirname, "../generated", name);
  const schemaDir = path.join(__dirname, "../schemas");

  beforeAll(async () => {
    if (fs.existsSync(generatedDir))
      fs.rmdirSync(generatedDir, { recursive: true });
    await generate({
      schemaFile: path.join(schemaDir, "format-schema.js"),
      schemaType: "custom",
      directory: generatedDir,
      addFormats: true,
    });
  });

  test("files should match", () => {
    const files = fs.readdirSync(generatedDir);
    expect(files).toEqual([
      "decoders.ts",
      "helpers.ts",
      "meta.ts",
      "models.ts",
      "schema.json",
      "validate.ts",
    ]);
  });

  test("schema should match", async () => {
    const file = fs.readFileSync(
      path.join(generatedDir, `schema.json`),
      "utf8"
    );
    expect(file).not.toBeUndefined();
    expect(file).toMatchSnapshot();
    expect(await new Ajv().validateSchema(JSON.parse(file))).toEqual(true);
  });

  test("decoders should match", () => {
    const file = fs.readFileSync(
      path.join(generatedDir, `decoders.ts`),
      "utf8"
    );
    expect(file).not.toBeUndefined();
    expect(file).toMatchSnapshot();
  });

  test("models should match", () => {
    const file = fs.readFileSync(path.join(generatedDir, `models.ts`), "utf8");
    expect(file).not.toBeUndefined();
    expect(file).toMatchSnapshot();
  });
});

describe("format-schema - compile based - options", () => {
  const name = "format";
  const generatedDir = path.join(__dirname, "../generated", `${name}-options`);
  const schemaDir = path.join(__dirname, "../schemas");

  beforeAll(async () => {
    if (fs.existsSync(generatedDir))
      fs.rmdirSync(generatedDir, { recursive: true });
    await generate({
      schemaFile: path.join(schemaDir, "format-schema.js"),
      schemaType: "custom",
      directory: generatedDir,
      addFormats: true,
      formatOptions: { mode: "fast", formats: ["date", "time"] },
    });
  });

  test("decoders should match", () => {
    const file = fs.readFileSync(
      path.join(generatedDir, `decoders.ts`),
      "utf8"
    );
    expect(file).not.toBeUndefined();
    expect(file).toMatchSnapshot();
  });
});

describe("format-schema - standalone", () => {
  const name = "format";
  const generatedDir = path.join(
    __dirname,
    "../generated",
    `${name}-standalone`
  );
  const schemaDir = path.join(__dirname, "../schemas");

  beforeAll(async () => {
    if (fs.existsSync(generatedDir))
      fs.rmdirSync(generatedDir, { recursive: true });
    await generate({
      schemaFile: path.join(schemaDir, "format-schema.js"),
      schemaType: "custom",
      directory: generatedDir,
      addFormats: true,
      standalone: { validatorOutput: "commonjs" },
    });
  });

  test("files should match", () => {
    const files = fs.readdirSync(generatedDir);
    expect(files).toEqual([
      "decoders",
      "helpers.ts",
      "meta.ts",
      "models.ts",
      "schema.json",
      "validate.ts",
    ]);
  });

  describe("User validator", () => {
    test("validator should contain properties", () => {
      const file = fs.readFileSync(
        path.join(generatedDir, `decoders/User/validator.js`),
        "utf8"
      );
      expect(file).not.toBeUndefined();
      expect(file).toContain(
        `require("ajv-formats/dist/formats").fullFormats.date`
      );
    });
  });
});
