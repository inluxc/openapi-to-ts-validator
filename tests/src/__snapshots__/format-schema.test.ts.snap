// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format-schema - compile based - options decoders should match 1`] = `
"/* eslint-disable */

import Ajv from \"ajv\";
import addFormats from \"ajv-formats\";
import type { Decoder } from \"./helpers\";
import { validateJson } from \"./validate\";
import type { User, Price } from \"./models\";
import jsonSchema from \"./schema.json\" assert { type: "json" };

export const ajv = new Ajv({ strict: false });
addFormats(ajv, { mode: \"fast\", formats: [\"date\", \"time\"] });
ajv.compile(jsonSchema);

// Decoders
export const UserDecoder: Decoder<User> = {
  definitionName: \"User\",
  schemaRef: \"#/definitions/User\",

  decode(json: unknown): User {
    const schema = ajv.getSchema(UserDecoder.schemaRef);
    if (!schema) {
      throw new Error(\`Schema \${UserDecoder.definitionName} not found\`);
    }
    return validateJson(json, schema, UserDecoder.definitionName);
  },
};
export const PriceDecoder: Decoder<Price> = {
  definitionName: \"Price\",
  schemaRef: \"#/definitions/Price\",

  decode(json: unknown): Price {
    const schema = ajv.getSchema(PriceDecoder.schemaRef);
    if (!schema) {
      throw new Error(\`Schema \${PriceDecoder.definitionName} not found\`);
    }
    return validateJson(json, schema, PriceDecoder.definitionName);
  },
};
"
`;

exports[`format-schema - compile based decoders should match 1`] = `
"/* eslint-disable */

import Ajv from \"ajv\";
import addFormats from \"ajv-formats\";
import type { Decoder } from \"./helpers\";
import { validateJson } from \"./validate\";
import type { User, Price } from \"./models\";
import jsonSchema from \"./schema.json\" assert { type: \"json\" };

export const ajv = new Ajv({ strict: false });
addFormats(ajv, undefined);
ajv.compile(jsonSchema);

// Decoders
export const UserDecoder: Decoder<User> = {
  definitionName: \"User\",
  schemaRef: \"#/definitions/User\",

  decode(json: unknown): User {
    const schema = ajv.getSchema(UserDecoder.schemaRef);
    if (!schema) {
      throw new Error(\`Schema \${UserDecoder.definitionName} not found\`);
    }
    return validateJson(json, schema, UserDecoder.definitionName);
  },
};
export const PriceDecoder: Decoder<Price> = {
  definitionName: \"Price\",
  schemaRef: \"#/definitions/Price\",

  decode(json: unknown): Price {
    const schema = ajv.getSchema(PriceDecoder.schemaRef);
    if (!schema) {
      throw new Error(\`Schema \${PriceDecoder.definitionName} not found\`);
    }
    return validateJson(json, schema, PriceDecoder.definitionName);
  },
};
"
`;

exports[`format-schema - compile based models should match 1`] = `
"/* eslint-disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

export interface User {
  createdAt: string;
  email: string;
}
export interface Price {
  amount: number;
}
"
`;

exports[`format-schema - compile based schema should match 1`] = `
"{
  \"type\": \"object\",
  \"title\": \"Schema\",
  \"definitions\": {
    \"User\": {
      \"type\": \"object\",
      \"properties\": {
        \"createdAt\": {
          \"type\": \"string\",
          \"format\": \"date\",
          \"formatMinimum\": \"2016-02-06\",
          \"formatExclusiveMaximum\": \"2016-12-27\"
        },
        \"email\": {
          \"type\": \"string\",
          \"format\": \"email\"
        }
      },
      \"required\": [
        \"createdAt\",
        \"email\"
      ]
    },
    \"Price\": {
      \"type\": \"object\",
      \"properties\": {
        \"amount\": {
          \"type\": \"number\",
          \"minimum\": 0,
          \"maximum\": 999
        }
      },
      \"required\": [
        \"amount\"
      ]
    }
  },
  \"properties\": {
    \"User\": {
      \"$ref\": \"#/definitions/User\"
    },
    \"Price\": {
      \"$ref\": \"#/definitions/Price\"
    }
  }
}"
`;
